merlin/systems/dag/ops/hugectr.py

```diff
@@ -0,0 +1,300 @@
+            for path in hugectr_model_path.glob(dense_pattern)
+            if "opt" not in path.name
+        ][0]
+        sparse_pattern = "*_sparse_*.model"
+        sparse_paths = [
+            os.path.join(hugectr_model_path, path.name)
+            for path in hugectr_model_path.glob(sparse_pattern)
+            if "opt" not in path.name
+        ]
+
+        config_dict = dict()
+        config_dict["supportlonglong"] = True
+
+        data_layer = model_json["layers"][0]
+        sparse_layers = [
+            layer
+            for layer in model_json["layers"]
+            if layer["type"] == "DistributedSlotSparseEmbeddingHash"
+        ]
+        num_cat_columns = sum(x["slot_num"] for x in data_layer["sparse"])
+        vec_size = [x["sparse_embedding_hparam"]["embedding_vec_size"] for x in sparse_layers]
+
+        model = dict()
+        model["model"] = model_name
+        model["slot_num"] = num_cat_columns
+        model["sparse_files"] = sparse_paths
+        model["dense_file"] = dense_path
+        model["maxnum_des_feature_per_sample"] = data_layer["dense"]["dense_dim"]
+        model["network_file"] = network_file
+        model["num_of_worker_buffer_in_pool"] = 4
+        model["num_of_refresher_buffer_in_pool"] = 1
+        model["deployed_device_list"] = self.device_list
+        model["max_batch_size"] = self.max_batch_size
+        model["default_value_for_each_table"] = [0.0] * len(sparse_layers)
+        model["hit_rate_threshold"] = 0.9
+        model["gpucacheper"] = self.hugectr_params["gpucacheper"]
+        model["gpucache"] = True
+        model["cache_refresh_percentage_per_iteration"] = 0.2
+        model["maxnum_catfeature_query_per_table_per_sample"] = [
+            len(x["sparse_embedding_hparam"]["slot_size_array"]) for x in sparse_layers
+        ]
+        model["embedding_vecsize_per_table"] = vec_size
+        model["embedding_table_names"] = [x["top"] for x in sparse_layers]
+        config_dict["models"] = [model]
+
+        parameter_server_config_path = str(node_export_path.parent / "ps.json")
+        with open(parameter_server_config_path, "w") as f:
+            f.write(json.dumps(config_dict))
+
+        self.hugectr_params["config"] = network_file
+
+        # These are no longer required from hugectr_backend release 3.7
+        self.hugectr_params["cat_feature_num"] = num_cat_columns
+        self.hugectr_params["des_feature_num"] = data_layer["dense"]["dense_dim"]
+        self.hugectr_params["embedding_vector_size"] = vec_size[0]
+        self.hugectr_params["slots"] = num_cat_columns
+        self.hugectr_params["label_dim"] = data_layer["label"]["label_dim"]
+
+        config = _hugectr_config(node_name, self.hugectr_params, max_batch_size=self.max_batch_size)
+
+        with open(os.path.join(node_export_path, "config.pbtxt"), "w") as o:
+            text_format.PrintMessage(config, o)
+
+        return config
+
+
+def _hugectr_config(name, hugectr_params, max_batch_size=None):
```

Comment:
It seems like there's a fair amount of repetition in this method. Maybe some of this can be done with a for loop?

Reply:
I'm pretty sure we implemented this a for loop @jperez999  must have been during the splitting of commits from https://github.com/NVIDIA-Merlin/systems/pull/125